diff --git a/dlls/win32u/driver.c b/dlls/win32u/driver.c
index f3c1d419b1d..e09e56a8244 100644
--- a/dlls/win32u/driver.c
+++ b/dlls/win32u/driver.c
@@ -754,16 +754,7 @@ static BOOL nulldrv_CreateDesktopWindow( HWND hwnd )
 
 static BOOL nodrv_CreateWindow( HWND hwnd )
 {
-    static int warned;
-    HWND parent = NtUserGetAncestor( hwnd, GA_PARENT );
-
-    /* HWND_MESSAGE windows don't need a graphics driver */
-    if (!parent || parent == NtUserGetThreadInfo()->msg_window) return TRUE;
-    if (warned++) return FALSE;
-
-    ERR_(winediag)( "Application tried to create a window, but no driver could be loaded.\n" );
-    if (driver_load_error[0]) ERR_(winediag)( "%s\n", debugstr_w(driver_load_error) );
-    return FALSE;
+    return TRUE;
 }
 
 static BOOL nulldrv_CreateWindow( HWND hwnd )
@@ -841,7 +832,7 @@ static void nulldrv_SetWindowText( HWND hwnd, LPCWSTR text )
 
 static UINT nulldrv_ShowWindow( HWND hwnd, INT cmd, RECT *rect, UINT swp )
 {
-    return ~0; /* use default implementation */
+    return 0;
 }
 
 static LRESULT nulldrv_SysCommand( HWND hwnd, WPARAM wparam, LPARAM lparam )
