diff --git a/dlls/shell32/shelllink.c b/dlls/shell32/shelllink.c
index 656357bf0a9..1e8c8de7920 100644
--- a/dlls/shell32/shelllink.c
+++ b/dlls/shell32/shelllink.c
@@ -299,43 +299,7 @@ static HRESULT WINAPI IPersistFile_fnLoad(IPersistFile* iface, LPCOLESTR pszFile
 
 BOOL run_winemenubuilder( const WCHAR *args )
 {
-    LONG len;
-    LPWSTR buffer;
-    STARTUPINFOW si;
-    PROCESS_INFORMATION pi;
-    BOOL ret;
-    WCHAR app[MAX_PATH];
-    void *redir;
-
-    GetSystemDirectoryW( app, MAX_PATH );
-    lstrcatW( app, L"\\winemenubuilder.exe" );
-
-    len = (lstrlenW( app ) + lstrlenW( args ) + 1) * sizeof(WCHAR);
-    buffer = heap_alloc( len );
-    if( !buffer )
-        return FALSE;
-
-    lstrcpyW( buffer, app );
-    lstrcatW( buffer, args );
-
-    TRACE("starting %s\n",debugstr_w(buffer));
-
-    memset(&si, 0, sizeof(si));
-    si.cb = sizeof(si);
-
-    Wow64DisableWow64FsRedirection( &redir );
-    ret = CreateProcessW( app, buffer, NULL, NULL, FALSE, DETACHED_PROCESS, NULL, NULL, &si, &pi );
-    Wow64RevertWow64FsRedirection( redir );
-
-    heap_free( buffer );
-
-    if (ret)
-    {
-        CloseHandle( pi.hProcess );
-        CloseHandle( pi.hThread );
-    }
-
-    return ret;
+    return TRUE;
 }
 
 static BOOL StartLinkProcessor( LPCOLESTR szLink )
